src/webgpu/shader/validation/decl/ptr_spelling.spec.ts

```diff
@@ -0,0 +1,157 @@
+export const description = `
+Validate spelling of pointer types.
+
+Pointer types may appear.
+
+They are parameterized by:
+- address space, always
+- store type
+- and access mode, as specified by the table in Address Spaces.
+   Concretely, only 'storage' address space allows it, and allows 'read', and 'read_write'.
+
+A pointer type can be spelled only if it corresponds to a variable that could be
+declared in the program.  So we need to test combinations against possible variable
+declarations.
+`;
+
+// This file tests spelling of the pointer type on let-declared pointers.
+//
+// Spelling of pointer-typed parameters on user-declared functions is tested by
+// webgpu:shader,validation,functions,restrictions:function_parameter_types:"*"
+
+import { makeTestGroup } from '../../../../common/framework/test_group.js';
+import { keysOf } from '../../../../common/util/data_tables.js';
+import { AddressSpace, kAccessModeInfo, kAddressSpaceInfo } from '../../types.js';
+import { ShaderValidationTest } from '../shader_validation_test.js';
+
+import {
+  pointerType,
+  infoExpander,
+  explicitSpaceExpander,
+  accessModeExpander,
+  getVarDeclShader,
+  supportsWrite,
+  ShaderStage,
+} from './util.js';
+
+// Address spaces that can hold an i32 variable.
+const kNonHandleAddressSpaces = keysOf(kAddressSpaceInfo).filter(
+  as => as !== 'handle'
+) as AddressSpace[];
+
+export const g = makeTestGroup(ShaderValidationTest);
+
+g.test('let_ptr_explicit_type_matches_var')
+  .desc(
+    'Let-declared pointer with explicit type initialized from var with same address space and access mode'
+  )
+  .specURL('https://w3.org/TR#ref-ptr-types')
+  .params(u =>
+    u // Generate non-handle variables in all valid permutations of address space and access mode.
+      .combine('addressSpace', kNonHandleAddressSpaces)
+      .expand('info', infoExpander)
```

Comment:
Maybe I'm misunderstanding things, but it looks like you're using the test parameters as a scratch location for putting intermediate state used by other `expand()` and `combine()` transforms.

Parameters should be just that - things that appear in the case query. There's a limit to the length of the query string, I think our CTS runner will fail at 250-ish characters. You should also consider that these are things that people type, so we want to avoid redundant information.

As far as I understand, `info` is immutable data fetched by looking up `kAddressSpaceInfo` with `addressSpace`. Given that this deterministically derived from `addressSpace`, I imagine you can push this lookup into `explicitSpaceExpander` and `accessModeExpander`, and drop `info` from the parameters.

I do believe that parameters with a leading underscore are excluded from the query, but I have no idea how efficient it is to do what you're doing.

Reply:
Fixed now.  Thanks.  Yep, it was dodgy.